src/components/ingestion/AbstractIngestionEngine.js

```diff
@@ -5,61 +5,90 @@ import {
 	AbstractInstantiationError,
 	NotImplementedError,
 } from '@tvkitchen/base-errors'
+import { TSDemuxer } from 'ts-demuxer'
+
 import logger from '%src/lib/logger'
 import kafka from '%src/lib/kafka'
+import { tsToSeconds } from '%src/lib/utils/mpegts'
 
 class AbstractIngestionEngine {
 	constructor() {
 		if (this.constructor === AbstractIngestionEngine) {
 			throw new AbstractInstantiationError(this.constructor.name)
 		}
 		this.producer = kafka.producer()
+		this.mpegtsDemuxer = new TSDemuxer(this.processDemuxedPacket)
+		this.ffmpegProcess = null
+		this.currentPts = 0
 	}
 
 	/**
-	 * Ingests data into the TV Kitchen pipeline.
+	 * Ingests a mpegts payload into the TV Kitchen pipeline.
 	 *
-	 * @param  {Buffer} data The buffer of data to be ingested.
-	 * @return {Boolean}     Whether or not the buffer was successfully ingested.
+	 * @param  {Payload} data The payload to be ingested.
+	 * @return {Boolean}      Whether or not the payload was successfully ingested.
 	 */
-	ingestData = async (data) => {
-		const payload = new Payload()
-		payload.data = data
+	ingestPayload = async (payload) => {
+		if (!(payload instanceof Payload)) {
+			throw new Error('ingestPayload must be passed a Payload object')
+		}
 		await this.producer.send({
-			topic: dataTypes.STREAM.DATA,
+			topic: dataTypes.STREAM.CONTAINER,
 			messages: [{
-				value: data,
+				value: Payload.serialize(payload),
 			}],
 		})
 		return true
 	}
 
 	/**
-	 * Continuously ingest of a stream of data.
+	 * Updates ingestion state based on the latest mpegts packet data as
+	 * parsed by our demuxer.
 	 *
-	 * @param  {ReadableStream} readableStream The stream of data to be recorded.
-	 * @return {ReadableStream}                The input stream (for chaining)
+	 * @param  {Packet} packet A Packet object, as defined by the TSDemuxer project
 	 */
-	recordStream = (readableStream) => readableStream
-		.on('data', async (data) => {
-			readableStream.pause()
-			await this.ingestData(data)
-			readableStream.resume()
+	processDemuxedPacket = (packet) => {
+		this.currentPts = packet.pts
+	}
+
+	/**
+	 * Process a new chunk of data from an mpegts stream.
+	 * This data is used to update local state and ultimately pipe to kafka.
+	 *
+	 * @param  {Buffer} mpegtsData The latest arbitrary chunk of buffered mpegts data.
+	 */
+	processMpegtsData = (mpegtsData) => {
+		this.ffmpegProcess.stdout.pause()
+		this.mpegtsDemuxer.process(mpegtsData)
+		const payload = new Payload({
+			data: mpegtsData,
+			ptsTime: tsToSeconds(this.currentPts),
```

Comment:
Argh is this exposing a code smell...

Would it be clearer if it was something more like:

```
const packet = this.getMostRecentPacket()
const payload = ... {
  ...
  ptsTime: tsToSeconts(packet.pts)
}
```

Reply:
…yes I think so! Because reading back over this… I think what I would _assume_ happens here is that `		this.mpegtsDemuxer.process(mpegtsData)` actually changes/transforms `mpegtsData` itself (even though that's JavaScript-impossible, I think, or at least _very_ unidiomatic), something that's necessary for when we pass it into the payload with `data: mpegtsData`. I think folks would be surprised to find out that `this.currentPts`, instead, is the recipient of `this.mpegtsDemuxer.process(mpegtsData)`.

So yeah, I think rewriting it closer to what you have it would improve legibility.

Reply:
_Oh, you edited with lots more info; now I have to re-read!_

Reply:
Hey there, I haven't spent enough time with this to know for sure, but using an on data event and pausing the stream here may not the most clean and efficient way to accomplish throttling. It's possible there's some limiting constraint here that I'm not seeing, or experience that informs this choice, so _waves hands_ ignore me if so, but stream backpressure might be a better way to do throttling. If you want, I can take a look at what's involved in that, and whether the result would be in fact cleaner/more efficient either for this PR or a subsequent one after creating an issue.

Reply:
@reefdog great point -- and also while reading your note I realized that instead of `getMostRecentPacket()` it might be more like `getMostRecentPacketMetadata()` or something along those lines.

Indeed, those are not actual "packets" from an mpegts spec perspective, but `Packet` as defined deep inside `TSDemuxer`

Reply:
Called this stuff `getMostRecentDemuxedPacket` which I think is the most accurate to what it is.